env/env_posix.cc

```diff
@@ -406,18 +382,34 @@ class PosixEnv : public CompositeEnvWrapper {
   }
 
  private:
-  std::vector<ThreadPoolImpl> thread_pools_;
-  pthread_mutex_t mu_;
-  std::vector<pthread_t> threads_to_join_;
+  friend Env* Env::Default();
+  // Constructs the default Env, a singleton
+  PosixEnv();
+
+  // The below 4 members are only used by the default PosixEnv instance.
+  // Non-default instances simply maintain references to the backing
+  // members in te default instance
+  std::vector<ThreadPoolImpl> thread_pools_storage_;
+  pthread_mutex_t mu_storage_;
+  std::vector<pthread_t> threads_to_join_storage_;
+  bool allow_non_owner_access_storage_;
+
+  std::vector<ThreadPoolImpl>& thread_pools_;
+  pthread_mutex_t& mu_;
+  std::vector<pthread_t>& threads_to_join_;
   // If true, allow non owner read access for db files. Otherwise, non-owner
   //  has no access to db files.
-  bool allow_non_owner_access_;
+  bool& allow_non_owner_access_;
 };
 
 PosixEnv::PosixEnv()
-    : CompositeEnvWrapper(this, FileSystem::Default().get()),
-      thread_pools_(Priority::TOTAL),
-      allow_non_owner_access_(true) {
+    : CompositeEnvWrapper(this, FileSystem::Default()),
```

Comment:
Why doesn't PosixEnv just inherit from Env?  What is gained by the Wrapper?

Reply:
Backward compatibility. The old/to-be-deprecated Env APIs should still work for the time being.

Reply:
Can you explain the compatibility?  I do not see where that is coming from.

Reply:
```Env::New*File``` APIs should call the corresponding ```FileSystem::New*File``` APIs. Since the former are pure virtual functions in Env, CompositeEnvWrapper provides the necessary wrappers to redirect those calls to the underlying FileSystem. Eventually, WinEnv should also inherit from it.